crates/typst/src/eval/plugin.rs

```diff
@@ -0,0 +1,70 @@
+use std::sync::Mutex;
+
+use crate::diag::SourceResult;
+use crate::util::Bytes;
+use host_wasmi::PluginInstance;
+use typst::diag::At;
+
+#[derive(Debug)]
+pub struct Plugin(Mutex<PluginInstance>, Bytes);
+
+impl PartialEq for Plugin {
+    fn eq(&self, other: &Self) -> bool {
+        self.1 == other.1
+    }
+}
+
+impl std::hash::Hash for Plugin {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.1.hash(state);
+    }
+}
+
+impl crate::eval::Type for Plugin {
+    const TYPE_NAME: &'static str = "Plugin";
+}
+
+impl Plugin {
+    pub fn new(instance: PluginInstance, bytes: Bytes) -> Self {
+        Self(Mutex::new(instance), bytes)
+    }
+    pub fn call(
+        &self,
+        function: &str,
+        args: &mut typst::eval::Args,
+    ) -> SourceResult<typst::eval::Value> {
+        let span = args.span;
+        let plugin = &mut self.0.lock().unwrap();
+        let ty = plugin
+            .get_function(function)
+            .ok_or("Plugin doesn't have the method: {function}")
+            .at(span)?
+            .ty(plugin.get_store());
+        let arg_count = ty.params().len();
+        let mut str_args = vec![];
+        for k in 0..arg_count {
+            let arg = args
+                .eat::<typst::eval::Value>()?
+                .ok_or(format!("plugin methods takes {arg_count} args, {k} provided"))
+                .at(span)?
+                .cast::<String>()
+                .at(span)?;
+            str_args.push(arg);
+        }
+        let s = plugin
+            .call(
+                function,
+                str_args.iter().map(|x| x.as_str()).collect::<Vec<_>>().as_slice(),
+            )
+            .at(span)?;
+        Ok(typst::eval::Value::Str(s.into()))
+    }
+
+    pub fn has_function(&self, method: &str) -> bool {
+        self.0.lock().unwrap().has_function(method)
+    }
+
+    pub fn iter_func(&self) -> Vec<String> {
```

Comment:
This clones all the string and collects into a vector, might be a problem but I ran into can't send a reference to temporary value and I didn't find a better solution yet

Reply:
If it's not on the hot path it's not that bad. I'd propose using `EcoString` instead as it has inline storage and function names will probably fit into 15 bytes most of the time.

Reply:
done